From e13d3aff7eaf2ad522900df8ef2a183fdb588617 Mon Sep 17 00:00:00 2001
From: Daniel Martin <dmanlfc@gmail.com>
Date: Wed, 7 Oct 2026 16:39:37 +1000
Subject: [PATCH] tqftpserv: read ACK block numbers as unsigned bytes

With a signed char the low byte of an ACK block number sign extends
from block 128 onwards, the reader then computes a bogus next block
and reads of large files never complete.
---
 tqftpserv.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tqftpserv.c b/tqftpserv.c
index 5231243..160cb74 100644
--- a/tqftpserv.c
+++ b/tqftpserv.c
@@ -738,7 +738,7 @@ static int handle_reader(struct tftp_client *client)
 	struct sockaddr_qrtr sq;
 	uint16_t block;
 	uint16_t last;
-	char buf[128];
+	uint8_t buf[128];
 	socklen_t sl;
 	ssize_t len;
 	ssize_t n = 0;
@@ -769,7 +769,7 @@ static int handle_reader(struct tftp_client *client)
 		if (err == ERROR_END_OF_TRANSFER)
 			log_info("Reader received end of transfer from %d:%d\n", sq.sq_node, sq.sq_port);
 		else
-			log_err("Remote returned an error: %d - %s\n", err, buf + 4);
+			log_err("Remote returned an error: %d - %s\n", err, (char *)buf + 4);
 		return -1;
 	} else if (opcode != OP_ACK) {
 		log_err("Expected ACK, got %d\n", opcode);
-- 
2.55.0

